// electron-chrome-extension/lib/browser/fetcher/download-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
// @ts-ignore
const download_crx_1 = require("download-crx");
const tmp_promise_1 = require("tmp-promise");
const querystring_1 = require("querystring");
const location_1 = tslib_1.__importDefault(require("./location"));
let fetch;
// require electron main context
if (process.env.NODE_ENV !== 'test') {
    fetch = require('electron-fetch').default;
}
const chromeWebStoreVersion = process.versions.chrome.split('.').slice(0, 2).join('.');
class DownloadProvider {
    constructor() {
        this.downloads = new Map();
    }
    downloadById(extensionId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tempDir = yield tmp_promise_1.dir({ prefix: 'ecx-', unsafeCleanup: true });
            this.downloads.set(extensionId, tempDir);
            const path = yield download_crx_1.downloadById(extensionId, tempDir.path, extensionId);
            return {
                id: extensionId,
                location: new location_1.default(path),
            };
        });
    }
    cleanupById(extensionId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tmpDir = this.downloads.get(extensionId);
            if (tmpDir) {
                yield tmpDir.cleanup();
            }
        });
    }
    getUpdateInfo(extension) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { id, version, updateUrl } = extension;
            const extQuery = querystring_1.stringify({
                id,
                v: version.number,
                installsource: 'ondemand',
                uc: '',
            });
            const checkQuery = querystring_1.stringify({
                response: 'updatecheck',
                prodversion: chromeWebStoreVersion,
                x: [extQuery],
            });
            // @ts-ignore
            const response = yield fetch(`${updateUrl}?${checkQuery}`);
            if (!response.ok) {
                throw new Error(`Http Status not ok: ${response.status} ${response.statusText}`);
            }
            const xml = yield response.text();
            return {
                xml,
            };
        });
    }
}
exports.default = DownloadProvider;
