// electron-chrome-extension/lib/browser/fetcher/storage-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const electron_1 = require("electron");
const glob_1 = require("glob");
const path_1 = require("path");
const fs_extra_1 = require("fs-extra");
// @ts-ignore
const unzip_crx_3_1 = tslib_1.__importDefault(require("unzip-crx-3"));
const location_1 = tslib_1.__importDefault(require("./location"));
const defaultConfiguration = process.env.NODE_ENV !== 'test'
    ? {
        extensionsFolder: new location_1.default(path_1.join(electron_1.app.getPath('userData'), 'ChromeExtensions')),
        cacheFolder: new location_1.default(path_1.join(electron_1.app.getPath('userData'), 'ChromeExtensions-cache')),
    }
    : {
        extensionsFolder: new location_1.default(''),
        cacheFolder: new location_1.default(''),
    };
class StorageProvider {
    constructor(customConfiguration = {}) {
        const { extensionsFolder, cacheFolder, } = Object.assign({}, defaultConfiguration, customConfiguration);
        this.extensionsFolder = extensionsFolder;
        this.cacheFolder = cacheFolder;
    }
    installExtension(crxDownload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tempDestination = path_1.resolve(this.extensionsFolder.path, this.cacheFolder.path, crxDownload.id);
            yield fs_extra_1.ensureDir(tempDestination);
            yield this.unzipCrx(crxDownload.location, new location_1.default(tempDestination));
            const manifest = yield this.readManifest(new location_1.default(path_1.join(tempDestination, 'manifest.json')));
            const versionDestination = path_1.resolve(this.extensionsFolder.path, crxDownload.id, manifest.version);
            yield fs_extra_1.ensureDir(versionDestination);
            // todo: don't overwrite directly
            yield fs_extra_1.move(tempDestination, versionDestination, { overwrite: true });
            return {
                id: crxDownload.id,
                location: new location_1.default(versionDestination),
                manifest,
            };
        });
    }
    getInstalledExtension() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const extensionsInstallations = new Map();
            const folders = glob_1.sync(path_1.join(this.extensionsFolder.path, '**/manifest.json'));
            for (const manifestPath of folders) {
                const extensionTree = manifestPath.split('/');
                const version = extensionTree[extensionTree.length - 2];
                const extensionId = extensionTree[extensionTree.length - 3];
                if (!extensionsInstallations[extensionId]) {
                    extensionsInstallations.set(extensionId, new Map());
                }
                const manifest = yield this.readManifest(new location_1.default(manifestPath));
                extensionsInstallations
                    .get(extensionId)
                    .set(version, {
                    id: extensionId,
                    // todo: refacto the slice op
                    location: new location_1.default(manifestPath.slice(0, -14)),
                    manifest,
                });
            }
            return extensionsInstallations;
        });
    }
    readManifest(location) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs_extra_1.readJson(location.path);
        });
    }
    unzipCrx(crxPath, destination) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                yield unzip_crx_3_1.default(crxPath.path, destination.path);
                return true;
            }
            catch (err) {
                return false;
            }
        });
    }
}
exports.default = StorageProvider;
