// electron-chrome-extension/lib/common/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.backgroundPageProcessFlag = '--background-page';
var Engine;
(function (Engine) {
    Engine["OnCreateRenderer"] = "cx-engine-create-renderer";
    Engine["OnExtensionLoaded"] = "cx-engine-extension-loaded";
    Engine["OnExtensionUnloaded"] = "cx-engine-extension-unloaded";
    Engine["GetExtension"] = "cx-engine-get-extension";
})(Engine = exports.Engine || (exports.Engine = {}));
var Protocol;
(function (Protocol) {
    Protocol["Extension"] = "chrome-extension:";
    Protocol["ExtensionDefault"] = "chrome-extension:";
    Protocol["Chrome"] = "chrome:";
})(Protocol = exports.Protocol || (exports.Protocol = {}));
var Channel;
(function (Channel) {
    Channel["Handler"] = "cx-handler";
    Channel["Event"] = "cx-event";
    Channel["Property"] = "cx-property";
})(Channel = exports.Channel || (exports.Channel = {}));
var Api;
(function (Api) {
    Api["Windows"] = "windows";
    Api["Tabs"] = "tabs";
    Api["Cookies"] = "cookies";
})(Api = exports.Api || (exports.Api = {}));
exports.scope = (channel, api) => [channel, api].join('-');
exports.eventScope = (api, eventName) => [api, eventName].join('.');
exports.extensionScope = (channel, api, extensionId) => [channel, api, extensionId].join('-');
