// electron-chrome-extension/lib/common/apis/cookies.js
"use strict";
// chrome.cookies
// See https://developer.chrome.com/extensions/cookies
Object.defineProperty(exports, "__esModule", { value: true });
// Types
var SameSiteStatus;
(function (SameSiteStatus) {
    SameSiteStatus["NoRestriction"] = "no_restriction";
    SameSiteStatus["Lax"] = "lax";
    SameSiteStatus["Strict"] = "strict";
})(SameSiteStatus = exports.SameSiteStatus || (exports.SameSiteStatus = {}));
var OnChangedCause;
(function (OnChangedCause) {
    OnChangedCause["Evicted"] = "evicted";
    OnChangedCause["Expired"] = "expired";
    OnChangedCause["Explicit"] = "explicit";
    OnChangedCause["ExpiredOverwrite"] = "expired_overwrite";
    OnChangedCause["Overwrite"] = "overwrite";
})(OnChangedCause = exports.OnChangedCause || (exports.OnChangedCause = {}));
// Methods
var Methods;
(function (Methods) {
    Methods["Get"] = "get";
    Methods["GetAll"] = "get-all";
    Methods["Set"] = "set";
    Methods["Remove"] = "remove";
    Methods["GetAllCookieStores"] = "get-all-cookie-stores";
})(Methods = exports.Methods || (exports.Methods = {}));
// Events
var Events;
(function (Events) {
    Events["OnChanged"] = "onChanged";
})(Events = exports.Events || (exports.Events = {}));
