// electron-chrome-extension/lib/common/apis/windows.js
"use strict";
// chrome.windows
// See https://developer.chrome.com/extensions/windows
Object.defineProperty(exports, "__esModule", { value: true });
var Methods;
(function (Methods) {
    Methods["Get"] = "get";
    Methods["GetCurrent"] = "get-current";
    Methods["GetLastFocused"] = "get-last-focused";
    Methods["GetAll"] = "get-all";
    Methods["Create"] = "create";
    Methods["Update"] = "update";
    Methods["Remove"] = "remove";
})(Methods = exports.Methods || (exports.Methods = {}));
var WindowType;
(function (WindowType) {
    WindowType["Normal"] = "normal";
    WindowType["Popup"] = "popup";
    WindowType["DevTools"] = "devtools";
})(WindowType = exports.WindowType || (exports.WindowType = {}));
var WindowState;
(function (WindowState) {
    WindowState["Normal"] = "normal";
    WindowState["Minimized"] = "minimized";
    WindowState["Maximized"] = "maximized";
    WindowState["Fullscreen"] = "fullscreen";
})(WindowState = exports.WindowState || (exports.WindowState = {}));
var CreateType;
(function (CreateType) {
    CreateType["Normal"] = "normal";
    CreateType["Popup"] = "popup";
})(CreateType = exports.CreateType || (exports.CreateType = {}));
// Events
var Events;
(function (Events) {
    Events["OnCreated"] = "onCreated";
    Events["OnRemoved"] = "onRemoved";
})(Events = exports.Events || (exports.Events = {}));
