// electron-chrome-extension/lib/renderer/api/windows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const windows_1 = require("../../common/apis/windows");
const common_1 = require("../../common/");
const wire_1 = require("../wire");
const event_1 = tslib_1.__importDefault(require("./event"));
exports.windows = (extensionId) => {
    const bind = wire_1.createWire(common_1.Api.Windows, extensionId);
    return {
        WINDOW_ID_NONE: -1,
        WINDOW_ID_CURRENT: -2,
        get: bind(windows_1.Methods.Get),
        getCurrent: bind(windows_1.Methods.GetCurrent),
        getLastFocused: bind(windows_1.Methods.GetLastFocused),
        getAll: bind(windows_1.Methods.GetAll),
        create: bind(windows_1.Methods.Create),
        update: bind(windows_1.Methods.Update),
        remove: bind(windows_1.Methods.Remove),
        onCreated: new event_1.default(),
        onRemoved: new event_1.default(),
        onChanged: new event_1.default(),
        onFocusChanged: new event_1.default(),
    };
};
exports.default = exports.windows;
