// electron-chrome-extension/lib/browser/engine/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fetcher_1 = tslib_1.__importDefault(require("../fetcher"));
const chrome_extension_1 = require("../chrome-extension");
const types_1 = require("../../common/types");
class ECx {
    constructor() {
        this.loaded = new Map();
        this.needFirstScan = true;
    }
    stop() {
        fetcher_1.default.reset();
        this.unregisterExtensionUpdateListener();
    }
    setConfiguration(configuration = {}) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { fetcher, onUpdate } = configuration;
            this.stop();
            this.fetcher = new fetcher_1.default(fetcher);
            if (onUpdate) {
                this.registerExtensionUpdateListener(onUpdate);
            }
            return this;
        });
    }
    load(extensionId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.needFirstScan) {
                yield this.fetcher.scanInstalledExtensions();
                this.needFirstScan = false;
            }
            if (this.loaded.has(extensionId)) {
                return this.loaded.get(extensionId);
            }
            const extension = yield this.get(extensionId);
            const { location: { path } } = extension;
            chrome_extension_1.addExtension(extensionId, path);
            this.loaded.set(extensionId, extension);
            return extension;
        });
    }
    localLoad(name, version) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.loaded.has(name)) {
                return this.loaded.get(name);
            }
            const location = `./vendor/${name}/${version}`
            const extension = yield {
              id: name,
              location: location,
              version: version,
              updateUrl: ''
            }
            chrome_extension_1.addExtension(name, location);
            this.loaded.set(name, extension);
            console.log("extension", extension)
            return extension;
        });
    }
    unload(extensionId) {
        if (!this.loaded.has(extensionId)) {
            return;
        }
        chrome_extension_1.removeExtension(extensionId);
        this.loaded.delete(extensionId);
    }
    isLoaded(extensionId) {
        return this.loaded.has(extensionId);
    }
    isUpToDate(extensionId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return !(yield this.fetcher.checkForUpdate(extensionId));
        });
    }
    get(extensionId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const installedExtension = this.fetcher.get(extensionId);
            if (installedExtension) {
                const isUpToDate = yield this.isUpToDate(extensionId);
                if (isUpToDate) {
                    return installedExtension;
                }
                return yield this.fetcher.update(extensionId);
            }
            return yield this.fetcher.fetch(extensionId);
        });
    }
    sendEvent(event) {
        chrome_extension_1.sendEventToExtensions(event);
    }
    registerExtensionUpdateListener(callback) {
        this.unregisterExtensionUpdateListener();
        this.onExtensionUpdateListener = callback;
        this.fetcher.addListener(types_1.ExtensionStatus.Updated, this.onExtensionUpdateListener);
    }
    unregisterExtensionUpdateListener() {
        if (this.onExtensionUpdateListener) {
            this.fetcher.removeListener(types_1.ExtensionStatus.Updated, this.onExtensionUpdateListener);
        }
    }
}
exports.default = new ECx();
