// electron-chrome-extension/lib/browser/engine/handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = require("path");
const { RpcIpcManager } = require('electron-simple-rpc');
const common_1 = require("../../common");
/**
 * @name Handler
 * @abstract This abstract class provide the glue to be triggered
 * from the renderer via RPC and emit events. The handler prepare
 * a custom scope per extension id for the RPC Manager and
 * assign methods defined in the enum to the linked handler method.
 * @example see `src/browser/handlers/cookies.ts`
 * @param extensionId string
 * @param emitter Function - the event bus that will forward events to renderers
 */
class Handler {
    constructor(extensionId, emitter) {
        this.namespace = this.constructor.name;
        this.emitter = emitter;
        const definitions = path_1.join('..', '..', 'common', 'apis', this.namespace.toLowerCase());
        const { Methods } = require(definitions);
        const library = Object
            .keys(Methods)
            .reduce((handlers, method) => {
            const handler = this[`handle${method}`];
            handlers[Methods[method]] = (...args) => tslib_1.__awaiter(this, void 0, void 0, function* () { return yield handler.apply(this, args); });
            return handlers;
        }, {});
        const handlerScope = common_1.extensionScope(common_1.Channel.Handler, common_1.Api[this.namespace], extensionId);
        this.manager = new RpcIpcManager(library, handlerScope);
    }
    emit(eventName, payload) {
        const channel = common_1.eventScope(common_1.Api[this.namespace], eventName);
        this.emitter({ channel, payload });
    }
    release() {
        this.manager.release();
    }
}
exports.default = Handler;
